### DISCUSSION TITLE
REPL predef functionality

### DISCUSSION THREAD

**Discussion by mpollmeier** [scala/scala3#17695] [Feature Requests]
The REPL currently doesn't have predef functionality, i.e. the ability to execute some code automatically on startup. I didn't find this being discussed or worked on. I happen to have that need, and this was [asked on Discord](https://discord.com/channels/632150470000902164/901153021638082660/938146035719610388) in the past. 

The below suits my needs, but I thought it might make sense to bring this into the standard repl, e.g. as a parameter, since it's a fairly common functionality of a repl. 

```scala
import dotty.tools.repl._
val repl = new ReplDriver(replArgs)
val predefCode =
  """
    |def foo = 42
    |import Int.MaxValue
    |""".stripMargin
val state = repl.run(predefCode)(repl.initialState)
repl.runUntilQuit(state)
```

For reference: Ammonite has are three distinct ways to pass predef code:
* `~/.ammonite/predef.sc` - runs every time this user starts ammonite
* `amm --predef-code 'def foo = 42'`
* `amm --predef someFile.sc`

Would you be open to a contribution that adds a predef functionality to the REPL, or is the intention to keep the repl as small as possible?

**Comment by som-snytt** [comment]
A `.scalarc`-style file is frequently requested.

Scala 2 has
```
 -i <file>    preload <file> before starting the REPL
 -I <file>    preload <file>, enforcing line-by-line interpretation
```
and
```
  -Yimports:<import>                    Custom root imports, default is `java.lang,scala,scala.Predef`.
```
It's worth distinguishing "predef" as "available definitions" from "initialization" as "things to evaluate on startup".

I don't know if there is a roadmap or policy for contributions, but there is also a useful distinction between "basic REPL that demonstrates language features" and "extended tool for prototyping or scripting". Ammonite and scala-cli serve the latter purpose.

**Comment by mpollmeier** [comment]
Ok, thanks for the feedback!

Your answer sounds like predef is probably not something we want to have in the standard scala3 repl - do I interpret that correctly? In that case we can close the ticket.

**Comment by som-snytt** [comment]
To clarify, I intended only to clarify terms. It seems to me a useful feature in some form.

> I don't know if there is a roadmap or policy for contributions
